base/array.jl

```diff
@@ -2331,20 +2331,46 @@ Int64[]
 function findall(A)
     collect(first(p) for p in pairs(A) if last(p))
 end
+
 # Allocating result upfront is faster (possible only when collection can be iterated twice)
-function findall(A::AbstractArray{Bool})
-    n = count(A)
+function findall(f::Function, A::AbstractArray{Bool})
+    # Compute f for true and false only once
+    ft, ff = f(true), f(false)
+    (ft | ff) || return Vector{eltype(keys(A))}()
+    (ft & ff) && return vec(Array(keys(A)))
+    n = let
+        c = count(A)
+        ft ? c : length(A) - c
+    end
     I = Vector{eltype(keys(A))}(undef, n)
+    _findall(ff, I, A)
+end
+
+function _findall(invert::Bool, I::Vector, A::AbstractArray{Bool})
     cnt = 1
-    for (i,a) in pairs(A)
-        if a
-            I[cnt] = i
-            cnt += 1
-        end
+    len = length(I)
+    @inbounds for (k, v) in pairs(A)
```

Comment:
this is probably not inbounds

Reply:
@vtjnash Why wouldn't this be in bounds? We known that `cnt` cannot be larger than `length(A)`.

Reply:
`Base` can't assume that `AbstractArray` have correct implementations of `length`

Reply:
@oscardssmith, I think we have some problems then :P

https://github.com/JuliaLang/julia/blob/5bf1ca5c724143091973ef6632f1b3026d8021af/base/sort.jl#L502-L517

https://github.com/JuliaLang/julia/blob/5bf1ca5c724143091973ef6632f1b3026d8021af/base/abstractarray.jl#L450-L453

https://github.com/JuliaLang/julia/blob/5bf1ca5c724143091973ef6632f1b3026d8021af/base/reduce.jl#L410-L431

etc etc

Reply:
But also `I` is a `Vector` and is not abstract.